### ISSUE TITLE
Image block: Blob is not always a temporary image, it could be an external final image from 3rd party implementation

### ISSUE THREAD

**Issue by jasperck** [WordPress/gutenberg#60688]
## What problem does this address?

Currently, core/image sees blob as a temporary image, whereas sometimes a blob could be a final image in 3rd party implementations, for example, we handle image upload and host it on our end and wanted to pass a blob to core/image to display it correctly, it won't work. Passing some kind of ID would work around the first condition but will then break the isExternalImage condition.

```js
/**
 * Is the URL a temporary blob URL? A blob URL is one that is used temporarily
 * while the image is being uploaded and will not have an id yet allocated.
 *
 * @param {number=} id  The id of the image.
 * @param {string=} url The url of the image.
 *
 * @return {boolean} Is the URL a Blob URL
 */
const isTemporaryImage = ( id, url ) => ! id && isBlobURL( url );

/**
 * Is the url for the image hosted externally. An externally hosted image has no
 * id and is not a blob url.
 *
 * @param {number=} id  The id of the image.
 * @param {string=} url The url of the image.
 *
 * @return {boolean} Is the url an externally hosted url?
 */
export const isExternalImage = ( id, url ) => url && ! id && ! isBlobURL( url );
```

I feel like we unblock Image block for this case, thoughts?

**Comment by youknowriad** [comment]
I think that's a good point. 

To be honest, checking the url (if it's blob or not) has always been a hack. The ideal scenario is a way to define "local state" for blocks when creating blocks (dragging images, transforming...). This is the main reasons the blob check exists.

I think a lot of people have thought about this in the past and we may have related issues about it but no final solution yet.

Maybe @noisysocks or @mcsf have more context to share.

**Comment by mcsf** [comment]
I think too much has happened over too many years, so I wouldn't rely on historical context too much. :) I think this is worth exploring, yes. I don't know what the best approach is, but in my mind there are few orthogonal things to think about:

* Somehow fixing the matter of local state, as Riad terms it. This, I think, is the long road. If we fixed this, then we could introduce the corresponding block migrations and then _maybe_ be in a better position to address these third-party needs.
* Instead of the above, introduce an escape hatch in the form of one or two new attributes (e.g. `isExternal` or `isExternallyManaged`), which obviously default to false and should be ignored by 99% of users.
* Regardless of the escape hatch, I think we have a ~fragile~ limited model of what is temporary and what is external. The Image block was built to fit WP's model, and as such when images are "external" it is assumed that they aren't managed at all; in other words, they are resources found out there on the Web, and this explains why there are affordances encouraging the user to sideload these images to the WP instance (for safekeeping, I suppose).
  * Where I am getting at is that we could discuss expectations of users like @jasperck to try to redefine notions like "external", "externally managed", etc.

**Comment by mcsf** [comment]
I hadn't seen the newly created #60691 when I posted my comment, but that definitely touches on my point about consolidating the model.

**Comment by noisysocks** [comment]
Yeah it's something I've looked into a few times but always placed back into the "too hard" basket.

My conceptual understanding of the issue is that blocks can already use `useState` for private state which is good but what's missing is a way of passing "constructor arguments" to the block so that this local state can be initialised to an external value.  In this specific case that's needed so that e.g. the drag and drop handler can create an image block with the (blob) image that was just dropped.
